vnext/JSI/Shared/ChakraObjectRef.h

```diff
@@ -0,0 +1,141 @@
+    const char *const chakraApiName = nullptr);
+
+// An shared_ptr like RAII Wrapper for JsRefs, which are references to objects
+// owned by the garbage collector. These include JsContextRef, JsValueRef, and
+// JsPropertyIdRef, etc. ChakraObjectRef ensures that JsAddRef and JsRelease
+// are called upon initialization and invalidation, respectively. It also allows
+// users to implicitly convert it into a JsRef. A ChakraObjectRef must only be
+// initialized once and invalidated once.
+class ChakraObjectRef {
+ public:
+  ChakraObjectRef() noexcept {}
+  inline explicit ChakraObjectRef(JsRef ref) {
+    Initialize(ref);
+  }
+
+  ChakraObjectRef(const ChakraObjectRef &original) noexcept;
+  ChakraObjectRef(ChakraObjectRef &&original) noexcept;
+
+  ChakraObjectRef &operator=(const ChakraObjectRef &rhs) noexcept;
+  ChakraObjectRef &operator=(ChakraObjectRef &&rhs) noexcept;
+
+  ~ChakraObjectRef() noexcept;
+
+  void Initialize(JsRef ref);
+  void Invalidate();
+
+  inline operator JsRef() const noexcept {
+    return m_ref;
+  }
+
+ private:
+  void swap(ChakraObjectRef &other);
+
+  enum class State { Uninitialized, Initialized, Invalidated };
+
+  JsRef m_ref = nullptr;
+  State m_state = State::Uninitialized;
+};
+
+// jsValue must be managing a JsValueRef.
+JsValueType GetValueType(const ChakraObjectRef &jsValue);
+
+// jsPropId must be managing a JsPropertyIdRef.
+JsPropertyIdType GetPropertyIdType(const ChakraObjectRef &jsPropId);
+
+// jsPropId must be managing a JsPropertyIdRef of type JsPropertyIdTypeString.
+std::wstring GetPropertyName(const ChakraObjectRef &jsPropId);
+
+// jsPropId must be managing a JsPropertyIdRef of type JsPropertyIdTypeSymbol.
+// Returns a ChakraObjectRef managing a JsValueRef pointing to a JS Symbol.
+ChakraObjectRef GetPropertySymbol(const ChakraObjectRef &jsPropId);
+
+// utf8 does not have to be null terminated.
+ChakraObjectRef GetPropertyId(const char *const utf8, size_t length);
+
+ChakraObjectRef GetPropertyId(const std::wstring &utf16);
+
+// jsString must be managing a JS string. The returned std::string/std::wstring
+// is UTF-8/UTF-16 encoded. These functions copy the JS  string buffer into the
+// returned std::string/std::wstring.
+std::string ToStdString(const ChakraObjectRef &jsString);
+std::wstring ToStdWstring(const ChakraObjectRef &jsString);
+
+// Returns a ChakraObjectRef managing a JS string. utf8 and utf16 do not have to
+// be null terminated and are copied to JS engine owned memory.
+ChakraObjectRef ToJsString(const char *const utf8, size_t length);
+ChakraObjectRef ToJsString(const wchar_t *const utf16, size_t length);
+
+// jsValue must be mananing a JsValueRef. Returns a ChakraObjectRef managing the
+// return value of the JS .toString function.
+ChakraObjectRef ToJsString(const ChakraObjectRef &jsValue);
+
+// Returns a ChakraObjectRef managing a JS number.
+ChakraObjectRef ToJsNumber(int num);
+
+// Returns a ChakraObjectRef managing a JS Object. This Object is backed by data
```

Comment:
>// [](start = 0, length = 2)

(nit) Not related specifically to this CR, but don't we want to use doxygen style comments? 
Eg:
//!

Reply:
I was not aware of this. I can switch to doxygen style comments.

---
In reply to: [336150329](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336150329) [](ancestors = 336150329)